ssl/quic/quic_channel.c

```diff
@@ -0,0 +1,1639 @@
+        goto err;
+
+    if (ossl_quic_wire_encode_transport_param_bytes(&wpkt, QUIC_TPARAM_INITIAL_SCID,
+                                                    NULL, 0) == NULL)
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_MAX_IDLE_TIMEOUT,
+                                                   ch->max_idle_timeout))
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_MAX_UDP_PAYLOAD_SIZE,
+                                                   QUIC_MIN_INITIAL_DGRAM_LEN))
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_ACTIVE_CONN_ID_LIMIT,
+                                                   4))
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_INITIAL_MAX_DATA,
+                                                   ossl_quic_rxfc_get_cwm(&ch->conn_rxfc)))
+        goto err;
+
+    /*
+     * We actually want the default CWM for a new RXFC, but here we just use
+     * stream0 as a representative specimen. TODO(QUIC): revisit this when we
+     * support multiple streams.
+     */
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_INITIAL_MAX_STREAM_DATA_BIDI_LOCAL,
+                                                   ossl_quic_rxfc_get_cwm(&ch->stream0->rxfc)))
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_INITIAL_MAX_STREAM_DATA_BIDI_REMOTE,
+                                                   ossl_quic_rxfc_get_cwm(&ch->stream0->rxfc)))
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_INITIAL_MAX_STREAM_DATA_UNI,
+                                                   ossl_quic_rxfc_get_cwm(&ch->stream0->rxfc)))
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_INITIAL_MAX_STREAMS_BIDI,
+                                                   0))
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_INITIAL_MAX_STREAMS_UNI,
+                                                   0))
+        goto err;
+
+    if (!WPACKET_get_total_written(&wpkt, &buf_len))
+        goto err;
+
+    ch->local_transport_params = (unsigned char *)buf_mem->data;
+    buf_mem->data = NULL;
+
+    if (!WPACKET_finish(&wpkt))
+        goto err;
+
+    wpkt_valid = 0;
+
+    if (!ossl_quic_dhs_set_transport_params(ch->dhs, ch->local_transport_params,
+                                            buf_len))
+        goto err;
+
+    ok = 1;
+err:
+    if (wpkt_valid)
+        WPACKET_cleanup(&wpkt);
+    BUF_MEM_free(buf_mem);
+    return ok;
+}
+
+/*
+ * QUIC Channel: Ticker-Mutator
+ * ============================
+ */
+
+/*
+ * The central ticker function called by the reactor. This does everything, or
+ * at least everything network I/O related. Best effort - not allowed to fail
+ * "loudly".
```

Comment:
Why is this not allowed to fail "loudly"? There are some things which might occur in the TLS code which would be fatal errors if they happen. In those cases I'd like to call `ERR_raise()`....but that sounds like it might be too "loud"!? But without it we won't know the root cause of a failure.

Reply:
The assumption is that the ticker is called arbitrarily at various points, so if it fails 'loudly' it will have implications for the exposed behaviour of public APIs which happen to call the ticker.

If the CSM needs to enter an error state (e.g. TLS alert), that should be an explicit state transition. That could happen in the ticker but it is not a failure of the ticker function per se.

Reply:
I think this is quite problematic. If we cannot call ERR_raise() then we are going to have a major problem resolving issues that occur, i.e. "somehow we entered into an error state but we have no idea how we got there or why".

> if it fails 'loudly' it will have implications for the exposed behaviour of public APIs which happen to call the ticker.

Don't they just fail?

> If the CSM needs to enter an error state (e.g. TLS alert), that should be an explicit state transition. That could happen in the ticker but it is not a failure of the ticker function per se.

I have no problem with the first sentence. I'm struggling to see why that last sentence should be the case.

Reply:
> I think this is quite problematic. If we cannot call ERR_raise() then we are going to have a major problem resolving issues that occur, i.e. "somehow we entered into an error state but we have no idea how we got there or why".

Could we just mark the file/line in the CSM and put that error on the stack only when we call APIs that are supposed to report an error if CSM is in the error state?